File: python-algorithm/ds_5_tree/tree_1_1_bool_recurrent/L101.py
#--coding:utf-8--

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        return self.isSym(root, root)

    def isSym(self, r1, r2):
        if not r1 and not r2:
            return True
        if not r1 or not r2:
            return False
        if r1.val != r2.val:
            return False
        return self.isSym(r1.left, r2.right) and self.isSym(r1.right, r2.left)

# bfs
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        q = collections.deque()
        q.append(root.left)
        q.append(root.right)
        while q:
            cur_nd1 = q.popleft()
            cur_nd2 = q.popleft()
            if not cur_nd1 and not cur_nd2:
                continue
            if not cur_nd1 or not cur_nd2:
                return False
            if cur_nd1.val != cur_nd2.val:
                return False
            q.append(cur_nd1.left)
            q.append(cur_nd2.right)
            q.append(cur_nd1.right)
            q.append(cur_nd2.left)
        return True

File: ds_5_tree/tree_1_1_bool_recurrent/L101.cpp


// 101 对称的二叉树

#include <iostream>
using namespace std;

//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


//1 递归写法
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (root == NULL) return true;
        return ismirror(root,root);
    }

    bool ismirror(TreeNode* t1,TreeNode* t2){
        if(t1 == NULL&&t2 == NULL) //都为空
            return true;
        if(t1 == NULL||t2 == NULL)//有一个为空
            return false;
        if(t1->val != t2->val)
            return false;
        return ismirror(t1->left,t2->right)&&ismirror(t1->right,t2->left);
    }
};

//2 非递归写法:队列迭代 类似BFS
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <queue>
class Solution1 {
public:
    bool isSymmetric(TreeNode* root) {
        if (root == NULL) return true;
        queue<TreeNode*> q;
        q.push(root->left);
        q.push(root->right);
        while(!q.empty()){
            TreeNode* lt = q.front();
            q.pop();
            TreeNode* rt = q.front();
            q.pop();
            if(lt == NULL && rt == NULL) continue;
            if(lt == NULL || rt == NULL) return false;
            if(lt->val != rt->val) return false;
            q.push(lt->left);
            q.push(rt->right);
            q.push(lt->right);
            q.push(rt->left);
        }
        return true;
    }
};